File: Android/app/src/main/java/com/gokuencinar/iruniversal/online/OnlineIrLibrary.kt
package com.gokuencinar.iruniversal.online

import com.gokuencinar.iruniversal.flipper.FlipperIrCodec
import com.gokuencinar.iruniversal.flipper.ImportedIrSignal
import com.gokuencinar.iruniversal.flipper.IrdbCsvCodec
import com.gokuencinar.iruniversal.ir.DeviceCategory
import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.net.HttpURLConnection
import java.net.URL
import java.util.Locale

enum class OnlineIrSource(val title: String) {
    FLIPPER_COMMUNITY("Flipper-IRDB"),
    FLIPPER_OFFICIAL("Flipper IRDB oficial"),
    LEGACY_IRDB("IRDB Web")
}

data class OnlineIrRemote(
    val id: String,
    val source: OnlineIrSource,
    val path: String,
    val brand: String,
    val model: String,
    val categoryLabel: String,
    val downloadUrl: String,
    val score: Int
) {
    val displayName: String
        get() = if (model.isBlank() || model.equals(brand, true)) brand else brand + " · " + model
}

data class OnlineLoadedRemote(
    val name: String,
    val sourceDescription: String,
    val signals: List<ImportedIrSignal>
)

class OnlineIrLibrary {
    private val textCache = mutableMapOf<String, String>()

    fun search(
        brand: String,
        model: String,
        category: DeviceCategory,
        sources: List<OnlineIrSource> = OnlineIrSource.entries,
        deep: Boolean = false
    ): List<OnlineIrRemote> {
        val b = clean(brand)
        val m = clean(model)
        if (b.isBlank() && m.isBlank()) return emptyList()

        val found = mutableListOf<OnlineIrRemote>()
        sources.forEach { source ->
            runCatching {
                when (source) {
                    OnlineIrSource.FLIPPER_COMMUNITY -> found += searchGitHubTree(
                        "https://api.github.com/repos/Lucaslhm/Flipper-IRDB/git/trees/main?recursive=1",
                        "https://raw.githubusercontent.com/Lucaslhm/Flipper-IRDB/main/",
                        source, b, m, category, deep
                    )
                    OnlineIrSource.FLIPPER_OFFICIAL -> found += searchGitHubTree(
                        "https://api.github.com/repos/flipperdevices/IRDB/git/trees/dev?recursive=1",
                        "https://raw.githubusercontent.com/flipperdevices/IRDB/dev/",
                        source, b, m, category, deep
                    )
                    OnlineIrSource.LEGACY_IRDB -> found += searchLegacy(b, m, category, deep)
                }
            }
        }

        return found
            .distinctBy { it.id }
            .sortedWith(compareByDescending<OnlineIrRemote> { it.score }.thenBy { it.displayName.lowercase() })
            .take(120)
    }

    fun download(remote: OnlineIrRemote): OnlineLoadedRemote {
        val text = fetchText(remote.downloadUrl, 2_000_000)
        val signals = if (remote.source == OnlineIrSource.LEGACY_IRDB) {
            IrdbCsvCodec.parse(text, "online-irdb")
        } else {
            FlipperIrCodec.parse(text)
        }
        require(signals.isNotEmpty()) { "El archivo no contiene señales compatibles." }
        return OnlineLoadedRemote(
            remote.displayName,
            remote.source.title + " · " + remote.path,
            signals
        )
    }

    fun importUrl(value: String): OnlineLoadedRemote {
        val url = value.trim()
        require(url.startsWith("https://")) { "Solo se admiten URLs HTTPS." }
        val text = fetchText(normalizedUrl(url), 2_000_000)
        var signals = FlipperIrCodec.parse(text)
        if (signals.isEmpty()) signals = IrdbCsvCodec.parse(text, "url-irdb")
        require(signals.isNotEmpty()) { "No se encontraron señales .ir o IRDB compatibles." }
        val name = URL(normalizedUrl(url)).path.substringAfterLast('/').substringBeforeLast('.')
            .replace("_", " ").ifBlank { "Mando importado" }
        return OnlineLoadedRemote(name, "URL · " + URL(url).host, signals)
    }

    private fun searchGitHubTree(
        treeUrl: String,
        rawPrefix: String,
        source: OnlineIrSource,
        brand: String,
        model: String,
        category: DeviceCategory,
        deep: Boolean
    ): List<OnlineIrRemote> {
        val root = JSONObject(fetchText(treeUrl, 20_000_000))
        if (root.optBoolean("truncated", false)) error("El índice de GitHub llegó truncado.")
        val tree = root.getJSONArray("tree")
        val out = mutableListOf<OnlineIrRemote>()

        for (i in 0 until tree.length()) {
            val entry = tree.getJSONObject(i)
            if (entry.optString("type") != "blob") continue
            val path = entry.optString("path")
            if (!path.lowercase().endsWith(".ir")) continue
            if (!categoryMatches(path, category, source)) continue

            val pieces = path.split("/")
            val filename = pieces.lastOrNull()?.removeSuffix(".ir") ?: "Mando"
            val guessedBrand = guessBrand(pieces, source)
            val score = matchScore(brand, model, guessedBrand, filename, path, deep)
            if (score <= 0) continue

            out += OnlineIrRemote(
                source.name + ":" + path,
                source,
                path,
                guessedBrand.ifBlank { "Desconocida" },
                filename.replace("_", " "),
                category.shortTitle,
                rawPrefix + path.replace(" ", "%20"),
                score
            )
        }
        return out
    }

    private fun searchLegacy(
        brand: String,
        model: String,
        category: DeviceCategory,
        deep: Boolean
    ): List<OnlineIrRemote> {
        val indexUrl = "https://cdn.jsdelivr.net/gh/probonopd/irdb@master/codes/index"
        return fetchText(indexUrl, 8_000_000).lineSequence().mapNotNull { raw ->
            val path = raw.trim()
            if (!path.lowercase().endsWith(".csv")) return@mapNotNull null
            val parts = path.split("/")
            if (parts.size < 3) return@mapNotNull null
            val candidateBrand = parts[0]
            val candidateModel = parts[1]
            if (!legacyCategoryMatches(candidateModel, category) && !deep) return@mapNotNull null

            val score = matchScore(
                brand, model, candidateBrand, candidateModel,
                candidateBrand + " " + candidateModel + " " + path, deep
            )
            if (score <= 0) return@mapNotNull null

            OnlineIrRemote(
                "legacy:" + path,
                OnlineIrSource.LEGACY_IRDB,
                path,
                candidateBrand,
                candidateModel,
                category.shortTitle,
                "https://cdn.jsdelivr.net/gh/probonopd/irdb@master/codes/" + path.replace(" ", "%20"),
                score
            )
        }.toList()
    }

    private fun guessBrand(pieces: List<String>, source: OnlineIrSource): String {
        if (pieces.size < 2) return ""
        if (source == OnlineIrSource.FLIPPER_COMMUNITY) {
            return (if (pieces.size >= 3) pieces[1] else pieces[0]).replace("_", " ")
        }
        val index = pieces.indexOf("categories")
        if (index >= 0 && pieces.size > index + 2) return pieces[index + 2].replace("_", " ")
        return pieces.dropLast(1).lastOrNull()?.replace("_", " ").orEmpty()
    }

    private fun categoryMatches(path: String, category: DeviceCategory, source: OnlineIrSource): Boolean {
        val n = clean(path)
        return when (category) {
            DeviceCategory.TELEVISION -> n.contains("tv") || n.contains("television")
            DeviceCategory.AIR_CONDITIONER ->
                source == OnlineIrSource.FLIPPER_COMMUNITY &&
                    (n.contains(" ac ") || n.contains("air conditioner") || n.contains("acs"))
            DeviceCategory.PROJECTOR -> n.contains("projector")
        }
    }

    private fun legacyCategoryMatches(value: String, category: DeviceCategory): Boolean {
        val n = clean(value)
        return when (category) {
            DeviceCategory.TELEVISION -> n.contains("tv") || n.contains("television")
            DeviceCategory.AIR_CONDITIONER -> n.contains("air") || n == "ac" || n.contains(" ac ")
            DeviceCategory.PROJECTOR -> n.contains("projector")
        }
    }

    private fun matchScore(
        brand: String,
        model: String,
        candidateBrand: String,
        candidateModel: String,
        full: String,
        deep: Boolean
    ): Int {
        val cb = clean(candidateBrand)
        val cm = clean(candidateModel)
        val all = clean(full)
        var score = 0

        if (brand.isNotBlank()) {
            score += when {
                cb == brand -> 35
                cb.contains(brand) || brand.contains(cb) -> 24
                compact(cb).contains(compact(brand)) -> 18
                deep && tokens(brand).all { all.contains(it) } -> 8
                else -> return 0
            }
        }

        if (model.isNotBlank()) {
            score += when {
                cm == model -> 40
                cm.contains(model) || all.contains(model) -> 28
                compact(all).contains(compact(model)) -> 20
                deep && tokens(model).all { all.contains(it) } -> 10
                else -> return 0
            }
        }
        return score
    }

    private fun clean(value: String): String =
        value.lowercase(Locale.ROOT)
            .replace("_", " ").replace("-", " ").replace(".", " ").replace("/", " ")
            .split(Regex("\\s+")).filter { it.isNotBlank() }.joinToString(" ")

    private fun compact(value: String): String = clean(value).filter { it.isLetterOrDigit() }
    private fun tokens(value: String): List<String> = clean(value).split(" ").filter { it.length >= 2 }

    private fun normalizedUrl(value: String): String {
        if (!value.startsWith("https://github.com/")) return value
        val url = URL(value)
        val pieces = url.path.trim('/').split("/")
        return if (pieces.size >= 5 && pieces[2] == "blob") {
            "https://raw.githubusercontent.com/" + pieces[0] + "/" + pieces[1] + "/" +
                pieces[3] + "/" + pieces.drop(4).joinToString("/")
        } else value
    }

    private fun fetchText(url: String, maxBytes: Int): String {
        textCache[url]?.let { return it }
        val connection = URL(url).openConnection() as HttpURLConnection
        connection.connectTimeout = 15_000
        connection.readTimeout = 25_000
        connection.setRequestProperty("User-Agent", "IR-Universal-Android/0.1")
        connection.instanceFollowRedirects = true

        try {
            val status = connection.responseCode
            require(status in 200..299) { "Servidor HTTP " + status }
            val out = ByteArrayOutputStream()
            connection.inputStream.use { input ->
                val buffer = ByteArray(16 * 1024)
                var total = 0
                while (true) {
                    val n = input.read(buffer)
                    if (n <= 0) break
                    total += n
                    require(total <= maxBytes) { "La respuesta es demasiado grande." }
                    out.write(buffer, 0, n)
                }
            }
            return out.toString(Charsets.UTF_8.name()).also { textCache[url] = it }
        } finally {
            connection.disconnect()
        }
    }
}

File: Sources/OnlineIRLibrary.swift
import Foundation

struct OnlineIRRemote: Identifiable, Hashable {
    let id: String
    let source: OnlineIRSource
    let path: String
    let brand: String
    let model: String
    let categoryLabel: String
    let downloadURL: URL
    let score: Int

    var displayName: String {
        model.isEmpty || model.caseInsensitiveCompare(brand) == .orderedSame
            ? brand
            : "\(brand) · \(model)"
    }
}

enum OnlineIRSource: String, CaseIterable, Identifiable, Hashable {
    case flipperCommunity
    case flipperOfficial
    case legacyIRDB

    var id: String { rawValue }

    var title: String {
        switch self {
        case .flipperCommunity: return "Flipper-IRDB"
        case .flipperOfficial: return "Flipper IRDB oficial"
        case .legacyIRDB: return "IRDB Web"
        }
    }
}

enum OnlineIRSourceFilter: String, CaseIterable, Identifiable {
    case all, flipper, official, irdb
    var id: String { rawValue }

    var title: String {
        switch self {
        case .all: return "Todas"
        case .flipper: return "Flipper"
        case .official: return "Oficial"
        case .irdb: return "IRDB"
        }
    }

    var sources: [OnlineIRSource] {
        switch self {
        case .all: return OnlineIRSource.allCases
        case .flipper: return [.flipperCommunity]
        case .official: return [.flipperOfficial]
        case .irdb: return [.legacyIRDB]
        }
    }
}

struct OnlineIRLoadedRemote: Identifiable {
    let id = UUID()
    let name: String
    let sourceDescription: String
    let signals: [ImportedIRSignal]
}

private struct GitHubIRTree: Decodable {
    struct Entry: Decodable {
        let path: String
        let type: String
    }
    let tree: [Entry]
    let truncated: Bool
}

@MainActor
final class OnlineIRLibrary: ObservableObject {
    @Published private(set) var results: [OnlineIRRemote] = []
    @Published private(set) var status = "Busca por marca y, si lo conoces, por modelo."
    @Published private(set) var isSearching = false
    @Published private(set) var brands: [String] = []
    @Published private(set) var isLoadingBrands = false
    @Published private(set) var brandStatus = "Cargando índice de marcas…"

    private var memoryCache: [String: Data] = [:]

    func loadBrands(
        category: IRDeviceCategory,
        filter: OnlineIRSourceFilter
    ) async {
        isLoadingBrands = true
        brandStatus = "Actualizando marcas…"

        var found: [String] = []
        var failures = 0

        for source in filter.sources {
            do {
                switch source {
                case .flipperCommunity:
                    found += try await brandsFromGitHubTree(
                        url: "https://api.github.com/repos/Lucaslhm/Flipper-IRDB/git/trees/main?recursive=1",
                        source: source,
                        category: category
                    )

                case .flipperOfficial:
                    found += try await brandsFromGitHubTree(
                        url: "https://api.github.com/repos/flipperdevices/IRDB/git/trees/dev?recursive=1",
                        source: source,
                        category: category
                    )

                case .legacyIRDB:
                    found += try await brandsFromLegacy(
                        category: category
                    )
                }
            } catch {
                failures += 1
            }
        }

        var unique: [String: String] = [:]

        for value in found {
            let display =
                value
                .replacingOccurrences(
                    of: "_",
                    with: " "
                )
                .trimmingCharacters(
                    in: .whitespacesAndNewlines
                )

            let key = clean(display)

            guard
                !key.isEmpty,
                key != "unknown",
                key != "desconocida"
            else {
                continue
            }

            if unique[key] == nil {
                unique[key] = display
            }
        }

        brands =
            unique.values.sorted {
                $0.localizedCaseInsensitiveCompare(
                    $1
                ) == .orderedAscending
            }

        isLoadingBrands = false

        if brands.isEmpty {
            brandStatus =
                failures > 0
                ? "No se pudo cargar el índice de marcas."
                : "No hay marcas para esta categoría."
        } else {
            brandStatus =
                "\(brands.count) marcas disponibles"
        }
    }

    func search(
        brand: String,
        model: String,
        category: IRDeviceCategory,
        filter: OnlineIRSourceFilter,
        deep: Bool
    ) async {
        let b = clean(brand)
        let m = clean(model)
        guard !b.isEmpty || !m.isEmpty else {
            results = []
            status = "Escribe al menos una marca o un modelo."
            return
        }

        isSearching = true
        status = "Consultando bibliotecas IR…"
        results = []

        var found: [OnlineIRRemote] = []
        var failures: [String] = []

        for source in filter.sources {
            do {
                switch source {
                case .flipperCommunity:
                    found += try await searchGitHubTree(
                        url: "https://api.github.com/repos/Lucaslhm/Flipper-IRDB/git/trees/main?recursive=1",
                        rawPrefix: "https://raw.githubusercontent.com/Lucaslhm/Flipper-IRDB/main/",
                        source: source,
                        brand: b,
                        model: m,
                        category: category,
                        deep: deep
                    )
                case .flipperOfficial:
                    found += try await searchGitHubTree(
                        url: "https://api.github.com/repos/flipperdevices/IRDB/git/trees/dev?recursive=1",
                        rawPrefix: "https://raw.githubusercontent.com/flipperdevices/IRDB/dev/",
                        source: source,
                        brand: b,
                        model: m,
                        category: category,
                        deep: deep
                    )
                case .legacyIRDB:
                    found += try await searchLegacy(
                        brand: b,
                        model: m,
                        category: category,
                        deep: deep
                    )
                }
            } catch {
                failures.append("\(source.title): \(error.localizedDescription)")
            }
        }

        var seen = Set<String>()
        results = found
            .sorted {
                $0.score == $1.score
                    ? $0.displayName.localizedCaseInsensitiveCompare($1.displayName) == .orderedAscending
                    : $0.score > $1.score
            }
            .filter { seen.insert($0.id).inserted }
            .prefix(120)
            .map { $0 }

        isSearching = false
        if results.isEmpty {
            status = failures.isEmpty
                ? "Sin resultados. Prueba «Búsqueda profunda» o deja el modelo vacío."
                : "Sin resultados. Algunas fuentes fallaron: \(failures.joined(separator: " · "))"
        } else {
            status = "Encontrados \(results.count) mandos/códigos. Se descargan solo cuando los abres."
        }
    }

    func download(_ remote: OnlineIRRemote) async throws -> OnlineIRLoadedRemote {
        let data = try await fetch(remote.downloadURL, maxBytes: 2_000_000)
        guard let text = String(data: data, encoding: .utf8) else {
            throw error("El archivo descargado no es texto UTF-8.")
        }

        let signals: [ImportedIRSignal]
        if remote.source == .legacyIRDB {
            signals = IRDBCSVCodec.parse(text: text, idPrefix: "online-irdb")
        } else {
            signals = FlipperIRCodec.parse(text: text)
        }

        guard !signals.isEmpty else {
            throw error("El archivo existe, pero no contiene señales compatibles.")
        }

        return OnlineIRLoadedRemote(
            name: remote.displayName,
            sourceDescription: "\(remote.source.title) · \(remote.path)",
            signals: signals
        )
    }

    func importURL(_ value: String) async throws -> OnlineIRLoadedRemote {
        let value = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let url = normalizedURL(value), url.scheme?.lowercased() == "https" else {
            throw error("Solo se admiten URLs HTTPS válidas.")
        }

        let data = try await fetch(url, maxBytes: 2_000_000)
        guard let text = String(data: data, encoding: .utf8) else {
            throw error("El contenido descargado no es texto UTF-8.")
        }

        var signals = FlipperIRCodec.parse(text: text)
        if signals.isEmpty {
            signals = IRDBCSVCodec.parse(text: text, idPrefix: "url-irdb")
        }
        guard !signals.isEmpty else {
            throw error("No se encontraron señales .ir o IRDB compatibles en esa URL.")
        }

        let base = url.deletingPathExtension().lastPathComponent
            .replacingOccurrences(of: "_", with: " ")
        return OnlineIRLoadedRemote(
            name: base.isEmpty ? "Mando importado" : base,
            sourceDescription: "URL · \(url.host ?? "HTTPS")",
            signals: signals
        )
    }

    private func brandsFromGitHubTree(
        url: String,
        source: OnlineIRSource,
        category: IRDeviceCategory
    ) async throws -> [String] {
        let data =
            try await cached(
                url,
                maxBytes: 20_000_000
            )

        let tree =
            try JSONDecoder().decode(
                GitHubIRTree.self,
                from: data
            )

        guard !tree.truncated else {
            throw error(
                "El índice de GitHub llegó incompleto."
            )
        }

        return tree.tree.compactMap {
            entry in

            guard
                entry.type == "blob",
                entry.path
                    .lowercased()
                    .hasSuffix(".ir"),
                categoryMatches(
                    entry.path,
                    category: category,
                    source: source
                )
            else {
                return nil
            }

            let pieces =
                entry.path
                    .split(separator: "/")
                    .map(String.init)

            let candidate =
                guessBrand(
                    pieces: pieces,
                    source: source
                )

            return candidate.isEmpty
                ? nil
                : candidate
        }
    }

    private func brandsFromLegacy(
        category: IRDeviceCategory
    ) async throws -> [String] {
        let indexURL =
            "https://cdn.jsdelivr.net/gh/probonopd/irdb@master/codes/index"

        let data =
            try await cached(
                indexURL,
                maxBytes: 8_000_000
            )

        guard
            let text =
                String(
                    data: data,
                    encoding: .utf8
                )
        else {
            return []
        }

        return text
            .components(
                separatedBy: .newlines
            )
            .compactMap { line in
                let path =
                    line.trimmingCharacters(
                        in:
                            .whitespacesAndNewlines
                    )

                guard
                    path.lowercased()
                        .hasSuffix(".csv")
                else {
                    return nil
                }

                let parts =
                    path
                        .split(separator: "/")
                        .map(String.init)

                guard parts.count >= 3 else {
                    return nil
                }

                let candidateBrand =
                    parts[0]
                let candidateModel =
                    parts[1]

                guard
                    legacyCategoryMatches(
                        candidateModel,
                        category: category
                    )
                else {
                    return nil
                }

                return candidateBrand
            }
    }

    private func searchGitHubTree(
        url: String,
        rawPrefix: String,
        source: OnlineIRSource,
        brand: String,
        model: String,
        category: IRDeviceCategory,
        deep: Bool
    ) async throws -> [OnlineIRRemote] {
        let data = try await cached(url, maxBytes: 20_000_000)
        let tree = try JSONDecoder().decode(GitHubIRTree.self, from: data)
        guard !tree.truncated else { throw error("El índice de GitHub llegó incompleto.") }

        return tree.tree.compactMap { entry in
            guard entry.type == "blob", entry.path.lowercased().hasSuffix(".ir") else { return nil }
            guard categoryMatches(entry.path, category: category, source: source) else { return nil }

            let pieces = entry.path.split(separator: "/").map(String.init)
            let file = pieces.last.map { String($0.dropLast(3)) } ?? "Mando"
            let guessedBrand = guessBrand(pieces: pieces, source: source)
            let score = matchScore(
                brand: brand,
                model: model,
                candidateBrand: guessedBrand,
                candidateModel: file,
                full: entry.path,
                deep: deep
            )
            guard score > 0 else { return nil }

            let encoded = entry.path.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? entry.path
            guard let downloadURL = URL(string: rawPrefix + encoded) else { return nil }

            return OnlineIRRemote(
                id: "\(source.rawValue):\(entry.path)",
                source: source,
                path: entry.path,
                brand: guessedBrand.isEmpty ? "Desconocida" : guessedBrand,
                model: file.replacingOccurrences(of: "_", with: " "),
                categoryLabel: category.shortTitle,
                downloadURL: downloadURL,
                score: score
            )
        }
    }

    private func searchLegacy(
        brand: String,
        model: String,
        category: IRDeviceCategory,
        deep: Bool
    ) async throws -> [OnlineIRRemote] {
        let indexURL = "https://cdn.jsdelivr.net/gh/probonopd/irdb@master/codes/index"
        let data = try await cached(indexURL, maxBytes: 8_000_000)
        guard let text = String(data: data, encoding: .utf8) else { return [] }

        return text.components(separatedBy: .newlines).compactMap { line in
            let path = line.trimmingCharacters(in: .whitespacesAndNewlines)
            guard path.lowercased().hasSuffix(".csv") else { return nil }
            let parts = path.split(separator: "/").map(String.init)
            guard parts.count >= 3 else { return nil }

            let candidateBrand = parts[0]
            let candidateModel = parts[1]
            let full = "\(candidateBrand) \(candidateModel) \(path)"
            if !legacyCategoryMatches(candidateModel, category: category) && !deep { return nil }

            let score = matchScore(
                brand: brand,
                model: model,
                candidateBrand: candidateBrand,
                candidateModel: candidateModel,
                full: full,
                deep: deep
            )
            guard score > 0 else { return nil }

            let encoded = path.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? path
            guard let u = URL(string: "https://cdn.jsdelivr.net/gh/probonopd/irdb@master/codes/" + encoded) else { return nil }

            return OnlineIRRemote(
                id: "legacy:\(path)",
                source: .legacyIRDB,
                path: path,
                brand: candidateBrand,
                model: candidateModel,
                categoryLabel: category.shortTitle,
                downloadURL: u,
                score: score
            )
        }
    }

    private func guessBrand(pieces: [String], source: OnlineIRSource) -> String {
        guard pieces.count >= 2 else { return "" }
        if source == .flipperCommunity {
            return pieces.count >= 3 ? pieces[1].replacingOccurrences(of: "_", with: " ") : pieces[0]
        }
        if let i = pieces.firstIndex(of: "categories"), pieces.count > i + 2 {
            return pieces[i + 2].replacingOccurrences(of: "_", with: " ")
        }
        return pieces.dropLast().last?.replacingOccurrences(of: "_", with: " ") ?? ""
    }

    private func categoryMatches(_ path: String, category: IRDeviceCategory, source: OnlineIRSource) -> Bool {
        let n = clean(path)
        switch category {
        case .television:
            return n.contains("tv") || n.contains("television")
        case .airConditioner:
            return source == .flipperCommunity && (n.contains("ac") || n.contains("air conditioner"))
        case .projector:
            return n.contains("projector")
        }
    }

    private func legacyCategoryMatches(_ value: String, category: IRDeviceCategory) -> Bool {
        let n = clean(value)
        switch category {
        case .television: return n.contains("tv") || n.contains("television")
        case .airConditioner: return n.contains("air") || n.contains("ac")
        case .projector: return n.contains("projector")
        }
    }

    private func matchScore(
        brand: String,
        model: String,
        candidateBrand: String,
        candidateModel: String,
        full: String,
        deep: Bool
    ) -> Int {
        let cb = clean(candidateBrand), cm = clean(candidateModel), all = clean(full)
        var score = 0

        if !brand.isEmpty {
            if cb == brand { score += 35 }
            else if cb.contains(brand) || brand.contains(cb) { score += 24 }
            else if compact(cb).contains(compact(brand)) { score += 18 }
            else if deep && tokens(brand).allSatisfy({ all.contains($0) }) { score += 8 }
            else { return 0 }
        }

        if !model.isEmpty {
            if cm == model { score += 40 }
            else if cm.contains(model) || all.contains(model) { score += 28 }
            else if compact(all).contains(compact(model)) { score += 20 }
            else if deep && tokens(model).allSatisfy({ all.contains($0) }) { score += 10 }
            else { return 0 }
        }
        return score
    }

    private func clean(_ value: String) -> String {
        value.folding(options: [.diacriticInsensitive, .caseInsensitive], locale: .current)
            .lowercased()
            .replacingOccurrences(of: "_", with: " ")
            .replacingOccurrences(of: "-", with: " ")
            .replacingOccurrences(of: ".", with: " ")
            .replacingOccurrences(of: "/", with: " ")
            .split(whereSeparator: { $0.isWhitespace })
            .joined(separator: " ")
    }

    private func compact(_ value: String) -> String {
        clean(value).filter { $0.isLetter || $0.isNumber }
    }

    private func tokens(_ value: String) -> [String] {
        clean(value).split(separator: " ").map(String.init).filter { $0.count >= 2 }
    }

    private func cached(_ string: String, maxBytes: Int) async throws -> Data {
        if let data = memoryCache[string] { return data }
        guard let url = URL(string: string) else { throw error("URL interna inválida.") }
        let data = try await fetch(url, maxBytes: maxBytes)
        memoryCache[string] = data
        return data
    }

    private func fetch(_ url: URL, maxBytes: Int) async throws -> Data {
        var request = URLRequest(url: url)
        request.timeoutInterval = 20
        request.setValue("IR-Universal/5.1 iOS", forHTTPHeaderField: "User-Agent")
        let (data, response) = try await URLSession.shared.data(for: request)
        if let http = response as? HTTPURLResponse, !(200...299).contains(http.statusCode) {
            throw error(http.statusCode == 403 || http.statusCode == 429
                ? "La fuente está limitando temporalmente las consultas."
                : "Servidor HTTP \(http.statusCode).")
        }
        guard data.count <= maxBytes else { throw error("La respuesta es demasiado grande.") }
        return data
    }

    private func normalizedURL(_ value: String) -> URL? {
        guard var c = URLComponents(string: value) else { return nil }
        if c.host?.lowercased() == "github.com" {
            let p = c.path.split(separator: "/").map(String.init)
            if p.count >= 5, p[2] == "blob" {
                let filePath = p.dropFirst(4).joined(separator: "/")
                return URL(string: "https://raw.githubusercontent.com/\(p[0])/\(p[1])/\(p[3])/\(filePath)")
            }
        }
        c.fragment = nil
        return c.url
    }

    private func error(_ message: String) -> NSError {
        NSError(domain: "IRUniversal.Online", code: 1, userInfo: [NSLocalizedDescriptionKey: message])
    }
}

enum IRDBCSVCodec {
    static func parse(text: String, idPrefix: String) -> [ImportedIRSignal] {
        let lines = text.components(separatedBy: .newlines)
        guard lines.count > 1 else { return [] }

        return lines.dropFirst().enumerated().compactMap { index, line in
            let f = csv(line)
            guard f.count >= 5,
                  let device = Int(f[2].trimmingCharacters(in: .whitespaces)),
                  let sub = Int(f[3].trimmingCharacters(in: .whitespaces)),
                  let function = Int(f[4].trimmingCharacters(in: .whitespaces)),
                  (0...255).contains(device), (0...255).contains(function)
            else { return nil }

            let proto = f[1].uppercased().replacingOccurrences(of: "-", with: "")
            let d = UInt8(device), command = UInt8(function)
            let id = "\(idPrefix):\(index):\(UUID().uuidString)"
            let code: IRCode?

            if proto.hasPrefix("NEC") {
                code = sub >= 0 && sub <= 255
                    ? IRProtocolEncoder.encode(protocolName: "NECEXT", address: [d, UInt8(sub)], command: [command, ~command], id: id)
                    : IRProtocolEncoder.encode(protocolName: "NEC", address: [d], command: [command], id: id)
            } else if proto.hasPrefix("RC5") {
                code = IRProtocolEncoder.encode(protocolName: "RC5", address: [d], command: [command], id: id)
            } else if proto.hasPrefix("RC6") {
                code = IRProtocolEncoder.encode(protocolName: "RC6", address: [d], command: [command], id: id)
            } else if proto.contains("SONY") || proto.contains("SIRC") {
                let p = proto.contains("20") ? "SIRC20" : proto.contains("15") ? "SIRC15" : "SIRC"
                code = IRProtocolEncoder.encode(protocolName: p, address: [d], command: [command], id: id)
            } else if proto == "JVC" || proto == "RCA" {
                code = IRProtocolEncoder.encode(protocolName: proto, address: [d], command: [command], id: id)
            } else {
                code = nil
            }

            guard let code else { return nil }
            let name = f[0].trimmingCharacters(in: .whitespacesAndNewlines)
            return ImportedIRSignal(
                name: name.isEmpty ? "BUTTON_\(function)" : name,
                code: code,
                sourceDescription: "IRDB \(proto)"
            )
        }
    }

    private static func csv(_ line: String) -> [String] {
        var out: [String] = [], cur = ""
        var quoted = false
        for ch in line {
            if ch == "\"" { quoted.toggle() }
            else if ch == "," && !quoted { out.append(cur); cur = "" }
            else { cur.append(ch) }
        }
        out.append(cur)
        return out
    }
}
